torch/sparse/triton_ops/triton_bsr_dense_mm.py

```diff
@@ -0,0 +1,473 @@
+
+    row_block_arange = tl.arange(0, BLOCKSIZE_ROW)
+    col_block_arange = tl.arange(0, BLOCKSIZE_COL)
+
+    # Pointers are set to the first block of the current row.
+    values_block_ptrs = (
+        values_ptr
+        + values_nnz_stride * row_idx_nnz_offset
+        + values_row_block_stride * row_block_arange[:, None]
+        + values_col_block_stride * col_block_arange[None, :]
+    )
+
+    # NOTE: dense is advanced into all dimensions but the tiled row one.
+    # That will be advanced in the loop according to values in col_indices.
+    dense_block_ptrs = (
+        dense_ptr
+        + dense_batch_stride * batch_idx
+        + dense_tiled_col_stride * col_block_pid
+        + dense_row_block_stride * col_block_arange[:, None]
+        + dense_col_block_stride * row_block_arange[None, :]
+    )
+
+    # Pointers are set to exact write-to locations
+    output_ptrs = (
+        output_ptr
+        + output_batch_stride * batch_idx
+        + output_tiled_row_stride * row_idx
+        + output_tiled_col_stride * col_block_pid
+        + output_row_block_stride * row_block_arange[:, None]
+        + output_col_block_stride * row_block_arange[None, :]
+    )
+
+    output_acc_block = tl.zeros((BLOCKSIZE_ROW, BLOCKSIZE_ROW), tl.float32)
+    col_index_nnz_ptr = col_indices_ptr + row_idx_nnz_offset * col_indices_stride
+    for _ in range(row_idx_nnz):
+        values_block = tl.load(values_block_ptrs)
+
+        # find which row of dense needs to get loaded
+        # for multiplication with values_block.
+        dense_row_idx = tl.load(col_index_nnz_ptr)
+        dense_block = tl.load(dense_block_ptrs + dense_tiled_row_stride * dense_row_idx)
+
+        # do block mm
+        output_acc_block += tl.dot(values_block, dense_block)
+
+        # move val/col_index ptrs to the next block in the row
+        values_block_ptrs += values_nnz_stride
+        col_index_nnz_ptr += col_indices_stride
+
+    # write back the result
+    tl.store(output_ptrs, output_acc_block.to(output_ptr.dtype.element_ty))
+
+
+def _run_sparse_rowspace_kernel(
+    blocksize, values, crow_indices, col_indices, dense, output
+):
+    # Compute a vector of non-zero elements numbers per each row.
+    # We want to ultimately iterate over non-zero rows.
+    nnz_per_row = crow_indices[:, 1:] - crow_indices[:, :-1]
+
+    # Compute indices of non-zero counts.
+    # batch_idx maps to a broadcasted batch index, while
+    # row_idx tracks non-zero rows of the sparse argument
+    # and rows of the output that get modified.
+    batch_idx, row_idx = nnz_per_row.nonzero(as_tuple=True)
+
+    # Compress the vector of counts to hold only non-zero values.
+    nnz_per_row = nnz_per_row[batch_idx, row_idx]
+    # Compute cumulative counts which along with nnz_per_row
+    # are used to compute offsets into nnz values.
+    nnz_per_row_cumsum = nnz_per_row.cumsum(-1)
+
+    # Launch kernel
+    n_nnz_block_rows = row_idx.size(-1)
+    n_block_cols = dense.size(-3)
```

Comment:
dense.size(-3) can easily be more than 65K

Reply:
My bad, will push this dim into the x-dim.

Reply:
Is there an easy way to launch the kernel in a loop in this case? That's what's usually done in eager.